### PR TITLE
Fix for tracked properties updating when using m3 native properties

### PR DESCRIPTION
**Pull request by igorT**
The fix in https://github.com/hjdivad/ember-m3/pull/1376 was not correct. In order to properly entangle we need to be doing `receiver.get` but to protect against re-entrance in production, we need to manually keep track of the key being accessed.

### CODE REVIEW

File: addon/model.js [hunk 1]
```diff
@@ -80,6 +80,10 @@ let megamorphicModelProxyHandler, megamorphicNativeDeprecationHandler;
 
 if (CUSTOM_MODEL_CLASS) {
   const MegamorphicModelProxyHandler = class {
+    constructor() {
+      this.getting = '';
+    }
```
Comment (rwjblue) [inline addon/model.js:85]: Let's use class fields instead (easier to reason about):

```suggestion
    this.getting = '';
```

**Review by rwjblue — COMMENTED** [review]
Given the explanation in the comments, I think there is probably a more straight forward solution for us.

Do we think (with the new tests added) that all of the cases that we care about are handled?

**Comment by igorT** [comment]
> Do we think (with the new tests added) that all of the cases that we care about are handled
I am not aware of any, do any come to mind?

File: addon/model.js [hunk 2]
```diff
@@ -88,19 +90,39 @@ if (CUSTOM_MODEL_CLASS) {
...
+      // of repeatadly calling `receiver.get`
+
+      if (!DEBUG) {
+        if (this.getting === key) {
+          this.getting = undefined;
+          return undefined;
+        }
+        this.getting = key;
+      }
+      return get(receiver, key);
```
Comment (rwjblue) [inline addon/model.js:125]:
```suggestion
     this.getting = key;
     try {
       return get(receiver, key);
     } finally {
       this.getting = undefined;
     }
```

File: tests/unit/model/native-access/tracked-properties-test.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        renderedTitle,
+        'Title: How to Win Friends and Influence People',
+        'component is rendered correctly'
+      );
+      book.title = 'New title';
+      await settled();
+
+      renderedTitle = this.element.querySelector('.title').innerText;
+      assert.equal(renderedTitle, 'Title: New title', 'component is rendered correctly');
+    });
```
Comment (rwjblue) [inline tests/unit/model/native-access/tracked-properties-test.js:76]: Can you add this exact test, but with the component:

```js
    test('Component args which depend on m3 native properties update correctly', async function (assert) {
      this.owner.register('service:m3-schema', TestSchema);
      this.owner.register(
        'template:components/show-book',
        hbs`<h1 class="title">{{@model.title}}</h1>
    `
      );

      this.store.push({
        data: {
          id: 'urn:li:book:1',
          type: 'com.example.bookstore.Book',
          attributes: {
            title: 'How to Win Friends and Influence People',
          },
        },
      });

      let book = (this.book = this.store.peekRecord('com.example.bookstore.Book', 'urn:li:book:1'));
      await render(hbs`
        {{show-book model=this.book}}
      `);

      let renderedTitle = this.element.querySelector('.title').innerText;
      assert.equal(
        renderedTitle,
        'Title: How to Win Friends and Influence People',
        'component is rendered correctly'
      );
      book.title = 'New title';
      await settled();

      renderedTitle = this.element.querySelector('.title').innerText;
      assert.equal(renderedTitle, 'Title: New title', 'component is rendered correctly');
    });
```

**Review by rwjblue — COMMENTED** [review]
Chatted a bit with @igorT about this (and my concerns in the prior comment). I think this is fine for now (with the tweak to use `try`/`finally` instead of relying on the nested access to reset), but we should work to refactor away from all of this.

The path forward is _basically_ to migrate to using `@glimmer/tracking/primitives/storage` (which allows us to remove the need for `Ember.get` in the proxy handler's `get` hook, by ensuring that our own tags are property consumed/entangled) via the https://github.com/ember-polyfills/ember-tracked-storage-polyfill. Once we do that then we can make our proxy return `undefined` for both `unknownProperty` and `setUknownProperty`.

Those changes should likely be follow on work after this lands.

File: tests/unit/model/native-access/tracked-properties-test.js (hunk 3 above)
Comment (igorT) [reply tests/unit/model/native-access/tracked-properties-test.js]: Added one manually

**Review by rwjblue — APPROVED** [review]

**Pull request merged**